spec/abstract/Structures_A.thy

```diff
@@ -384,13 +384,71 @@ datatype thread_state
   = Running
   | Inactive
   | Restart
-(*  | YieldTo obj_ref (* sc ref *)*)
   | BlockedOnReceive obj_ref "obj_ref option" receiver_payload
   | BlockedOnSend obj_ref sender_payload
   | BlockedOnReply obj_ref
   | BlockedOnNotification obj_ref
   | IdleThreadState
 
+(* FIXME RT: generating the following discriminators and selectors automatically breaks
+             unrelated proofs in strange ways *)
```

Comment:
How strange? I like the new consistency, but getting that from the datatype package would definitely be a bonus.

Reply:
It was before the break so I don't quite remember the details, but it was something to do with unification. Some proofs could be fixed by changing `force` to `auto`, but there was one in IpcDet_AI that I couldn't get working.

The strange part was that there were no obvious `thread_states` in the proof, and `apply_trace` didn't show any relevant rules being used.

Reply:
Once this is merged, we could try again with the updates you already have and compare what the datatype package generates to what you have here. Maybe that'll give us a hint.